add ad4170 device driver

buha opened the pull request:
Initial driver version with a device driver API that covers direct register access and higher level chip specific functionality.

## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

jansunil commented:
Hi  @buha - The same copy of the driver has been tested in the IIO FW. I confirm that this needs no mode changes and the version is stable for a release.

buha commented:
Cool, thanks for confirming @jansunil !

I pushed a new version which shows @jansunil and @mphalke as authors in the .c file, they contributed as well.

buha commented:
V2:
- fix attempt for doxygen errors
- renamed active_dev to id

buha commented:
CI fails building xilinx projects, it's unrelated.

amiclaus commented on drivers/adc/ad4170/ad4170.h:
```diff
@@ -0,0 +1,1142 @@
+/** ad4170_config default state when the chip is hardware/software reset. */
+extern struct ad4170_config ad4170_config_reset;
+
+int32_t ad4170_spi_reg_read(struct ad4170_dev *dev,
```
we should stick with our convention of using `int` type for functions.

buha replied:
will do

amiclaus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+	uint32_t reg;
+
+	ret = ad4170_spi_reg_write(dev, AD4170_REG_SCRATCH_PAD, pattern);
+	if (ret < 0)
```
same goes for `if (ret)` convention linux compliant.

buha replied:
will do

CiprianRegus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+
+NO_OS_DECLARE_CRC8_TABLE(ad4170_crc8);
+
+struct ad4170_config ad4170_config_reset = {
```
`static`

CiprianRegus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+#include "no_os_crc8.h"
+#include "ad4170.h"
+
+NO_OS_DECLARE_CRC8_TABLE(ad4170_crc8);
```
`static`

buha replied:
it is static: https://github.com/analogdevicesinc/no-OS/blob/main/include/no_os_crc8.h#L48

CiprianRegus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+ * 	CC BB AA FF EE DD 22 11 00
+ * etc.
+ */
+static void _memswap64(void * buf, uint32_t bytes, size_t step)
```
In the meantime, this function and `_is_bit_endian()` has been added to `no_os_util.c`, so we should use those instead.

CiprianRegus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+	}
+
+	*reg_data = 0;
+	if (dev->big_endian)
+		memcpy((uint8_t *)(reg_data + 1) - reg_size, &buf[di], reg_size);
+	else {
+		memcpy(reg_data, &buf[di], reg_size);
+		_memswap64(reg_data, reg_size, reg_size);
+	}
```
Can be replaced by `*reg_data = no_os_get_unaligned_be32()`? The endianess shouldn't matter in that case.

buha replied:
From memory, CPU endianess could be little or big, complicated by the fact that the data can come LSB or MSB first on the digital interface and this incoming data may not be aligned to powers of two. memswap64 has a step parameter which could be 3, as in the example. So I'm not touching this part, I'm pretty sure it's not possible to rewrite it using those macros.

CiprianRegus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+	if (!dev)
+		return -EINVAL;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, sizeof(buf));
```
return directly

buha replied:
will do

CiprianRegus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+	if (ret < 0)
+		return ret;
+
+	*data = (uint16_t)reg;
```
the cast is not needed

buha replied:
explicit casts carry meaning so i'll leave it as-is

CiprianRegus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2123 @@
+	if (ret < 0)
+		return ret;
+
+	*status = (uint8_t)reg;
```
In cases like these, I think we usually use `no_os_field_get()` along with some macros for the bit mask of the register field. I'm fine with either option.

buha commented:
V3:
- addressed some of the points of the reviews above

Coming up with an excuse here for not implementing complex changes - this driver was reviewed in 2021 and is known to work so I'll be implementing only trivial changes.

buha commented:
V4:
- reverted change of ad4170_config_reset to `static`, turns out it's prepended with `extern` in .h, I guess to publish the default device driver state after calling the reset API.

buha commented:
V5:
- astyle

danmois commented on drivers/adc/ad4170/ad4170.h:
```diff
@@ -0,0 +1,1142 @@
+/* AD4170_REG_INTERFACE_CONFIG_C */
+#define AD4170_INTERFACE_CONFIG_C_CRC_MSK			(NO_OS_BIT(7) | NO_OS_BIT(6) | NO_OS_BIT(1) | NO_OS_BIT(0))
+#define AD4170_INTERFACE_CONFIG_C_CRC(x)			(((~x) & 0x3) | (((x) << 6) & 0xC0))
+#define AD4170_INTERFACE_CONFIG_C_STRICT_REG_ACCESS_MSK	NO_OS_BIT(5)
```
Pay attention to white spaces and alignment.

buha replied:
will do

danmois commented on drivers/adc/ad4170/ad4170.h:
```diff
@@ -0,0 +1,1142 @@
+ * @brief ADC Operating Mode.
+ */
+enum ad4170_mode {
+	/** Continuous Conversion Mode. ADC converts continuously on the enabled channel(s) using Sinc-based filters. */
```
Should these comments be split on multiple lines?

buha replied:
hmmm, astyle didnt complain so i guess it's ok

danmois commented on drivers/adc/ad4170/ad4170.h:
```diff
@@ -0,0 +1,1142 @@
+ *   @brief  Header file for the ad4170 driver.
+ *   @author Darius Berghe (darius.berghe@analog.com)
+********************************************************************************
+ * Copyright 2020, 2023(c) Analog Devices, Inc.
```
2024?

buha replied:
these dates are correct from the point of view of when development was done, work is being published only now...  no idea which one should we pick

danmois commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2119 @@
+ *   @author Janani Sunil (janani.sunil@analog.com)
+ *   @author Mahesh Phalke (mahesh.phalke@analog.com)
+********************************************************************************
+ * Copyright 2020, 2023(c) Analog Devices, Inc.
```
2024?

buha commented:
V6:
- fixed macro alignment

danmois commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2119 @@
+	uint32_t addr = AD4170_ADDR(reg_addr);
+
+	/* max readable address in short instruction mode */
+	if (dev->spi_settings.short_instruction && addr > 0x3F)
```
Should we use macros for values like this one?

buha replied:
i'd leave it as-is because it has something to do with a certain spi standard and we don't have headers for that

danmois commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2119 @@
+	if (ret)
+		return ret;
+
+	/* TODO: set proper value, it's currently undefined in datasheet. */
```
Is there any datasheet available now?

buha replied:
good point, will try to find the right value

buha replied:
found it in the datasheet as Turn-On Settling Time, specified to 200us, so I will update accordingly

jansunil commented:
Hi All- I've tested the latest version here and things look ok.

jansunil commented on drivers/adc/ad4170/ad4170.h:
```diff
@@ -0,0 +1,1142 @@
+#define AD4170_REG_ADC_CTRL			(AD4170_R2B | 0x70)
+#define AD4170_REG_ERROR_EN			(AD4170_R2B | 0x72)
+#define AD4170_REG_ERROR			(AD4170_R2B | 0x74)
+#define AD4170_REG_INFO1			(AD4170_R2B | 0x76) /* TODO: implement this when it's specified in doc. */
```
Should we remove the TODO before release?

buha replied:
good point, turns out INFO1 is not in the latest datasheet, and is also unused in the driver so I'm removing it

jansunil commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2119 @@
+	if (ret)
+		return ret;
+
+	// TODO: handle 32b width. By default 24b data width is selected.
```
Should we remove the TODOs?

buha replied:
I'm not sure about this one, looks like there is some missing functionality but I can't add it without hardware setup, so I'd rather leave it in for now.

buha replied:
actually i re-read the datasheet and looks like continuous read works with 24b mode only so I guess we're good, I'm removing the TODO and the 32b handling that was commented out

buha commented:
V7:
- removed TODOs
- removed non existing register INFO1

buha commented:
I see it has been approved and tested, I will merge now

danmois reviewed (approved):
Looks good.

amiclaus commented on drivers/adc/ad4170/ad4170.h:
```diff
@@ -0,0 +1,1141 @@
+int ad4170_remove(struct ad4170_dev *dev);
+int ad4170_regmap(struct ad4170_dev *dev);
+#endif
+
```
nit: looks like multiple blank lines

amiclaus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2116 @@
+
+	return 0;
+}
+
```
ditto

amiclaus commented on drivers/adc/ad4170/ad4170.c:
```diff
@@ -0,0 +1,2116 @@
+	};
+
+	uint32_t ri;
+	printf("[register]: [value]\n");
```
let's go for `pr_info` (or even `pr_debug` in this particular case). same applies for all the other `printf` throughout the code.

amiclaus reviewed (commented):
some minor comments on my side.
